File: swift-package/brainflow/src/main/swift/brainflow/BoardShim.swift
//
//  BoardShim.swift
//  a Swift binding for BrainFlow's board_shim high-level API
//
//  Created by Scott Miller for Aeris Rising, LLC on 8/23/21.
//

import Foundation

/**
 * enable BrainFlow logger with level INFO
 */
func enableBoardLogger () {
    set_log_level (LogLevels.LEVEL_INFO.rawValue)
}

/**
 * enable BrainFlow logger with level TRACE
 */
func enableDevBoardLogger () {
    set_log_level (LogLevels.LEVEL_TRACE.rawValue)
}

/**
* disable BrainFlow logger
*/
func disableBoardLogger () {
    set_log_level (LogLevels.LEVEL_OFF.rawValue)
}

/**
 * redirect logger from stderr to a file
 */
func setLogFile (_ logFile: String) throws
{
    var cLogFile = logFile.cString(using: String.Encoding.utf8)!
    let result = set_log_file(&cLogFile)
    try checkErrorCode(errorMsg: "failed to set log file", errorCode: result)
}

/**
 * set log level
 */
func setLogLevel (_ logLevel: LogLevels) throws
{
    let result = set_log_level (logLevel.rawValue)
    try checkErrorCode(errorMsg: "failed to set log level", errorCode: result)
}

/**
 * send user defined strings to BrainFlow logger
 */
func logMessage (logLevel: Int32, message: String) throws {
    var cMessage = message.cString(using: String.Encoding.utf8)!
    let result = log_message (logLevel, &cMessage)
    try checkErrorCode(errorMsg: "Error in log_message", errorCode: result)
}

/**
 * get sampling rate for this board
 */
func getSamplingRate (boardId: BoardIds) throws -> Int32 {
    var samplingRate: Int32 = 0
    let result = get_sampling_rate (boardId.rawValue, &samplingRate)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return samplingRate
}

/**
 * get row index in returned by get_board_data() 2d array which contains timestamps
 */
func getTimestampChannel (boardId: BoardIds) throws -> Int32 {
    var timestampChannel: Int32 = 0
    let result = get_timestamp_channel (boardId.rawValue, &timestampChannel)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return timestampChannel
}

/**
 * get row index in returned by get_board_data() 2d array which contains markers
 */
func getMarkerChannel (boardId: BoardIds) throws -> Int32
{
    var markerChannel: Int32 = 0
    let result = get_marker_channel (boardId.rawValue, &markerChannel)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return markerChannel
}

/**
 * get number of rows in returned by get_board_data() 2d array
 */
func getNumRows (boardId: BoardIds) throws -> Int32 {
    var numRows: Int32 = 0
    let result = get_num_rows (boardId.rawValue, &numRows)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return numRows
}

/**
 * get row index in returned by get_board_data() 2d array which contains package nums
 */
func getPackageNumChannel (boardId: BoardIds) throws -> Int32 {
    var pkgNumChannel: Int32 = 0
    let result = get_package_num_channel (boardId.rawValue, &pkgNumChannel)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return pkgNumChannel
}

/**
 * get row index in returned by get_board_data() 2d array which contains battery level
 */
func getBatteryChannel (boardId: BoardIds) throws -> Int32 {
    var batteryChannel: Int32 = 0
    let result = get_battery_channel (boardId.rawValue, &batteryChannel)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return batteryChannel
}

/**
 * Get names of EEG electrodes in 10-20 system. Only if electrodes have freezed
 * locations
 */
func getEEGnames (boardId: BoardIds) throws -> [String] {
    var stringLen: Int32 = 0
    var bytes = [CChar](repeating: 0, count: 4096)
    let result = get_eeg_names (boardId.rawValue, &bytes, &stringLen)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)
    let EEGnames = bytes.toString(stringLen)

    return EEGnames.components(separatedBy: ",")
}

/**
 * Get board description
 */
func getBoardDescr (boardId: BoardIds) throws -> String {
    var boardDescrStr = [CChar](repeating: CChar(0), count: 16000)
    var stringLen: Int32 = 0
    let result = get_board_descr (boardId.rawValue, &boardDescrStr, &stringLen)
    try checkErrorCode(errorMsg: "failed to get board info", errorCode: result)

    if let description = String(data: Data(bytes: &boardDescrStr, count: Int(stringLen)), encoding: .utf8) {
        return description }
    else {
        return "no data found"
    }
}

/**
 * Get device name
 */
func getDeviceName (boardId: BoardIds) throws -> String {
    var stringLen: Int32 = 0
    var deviceName = [CChar](repeating: CChar(0), count: 4096)
    let result = get_device_name (boardId.rawValue, &deviceName, &stringLen)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return deviceName.toString(stringLen)
}

/**
 * get row indices in returned by get_board_data() 2d array which contain EEG
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getEEGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_eeg_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain EMG
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getEMGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_emg_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain ECG
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getECGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_ecg_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain
 * temperature data
 */
func getTemperatureChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_temperature_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain
 * resistance data
 */
func getResistanceChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_resistance_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain EOG
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getEOGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_eog_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain EXG
 * data
 */
func getEXGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_exg_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain EDA
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getEDAchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_eda_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain PPG
 * data, for some boards we can not split EEG\EMG\... and return the same array
 */
func getPPGchannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_ppg_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain accel
 * data
 */
func getAccelChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_accel_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain analog
 * data
 */
func getAnalogChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_analog_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
  * get row indices in returned by get_board_data() 2d array which contain gyro
  * data
  */
func getGyroChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_gyro_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/**
 * get row indices in returned by get_board_data() 2d array which contain other
 * data
 */
func getOtherChannels (boardId: BoardIds) throws -> [Int32] {
    var len: Int32 = 0
    var channels = [Int32](repeating: 0, count: 512)
    let result = get_other_channels (boardId.rawValue, &channels, &len)
    try checkErrorCode(errorMsg: "Error in board info getter", errorCode: result)

    return Array(channels[0..<Int(len)])
}

/*
 If the board is streaming or playback, then get the master board ID from params.other_info.
 Otherwise return the board ID itself.
 */
func getMasterBoardID(boardId: BoardIds, params: BrainFlowInputParams) throws -> BoardIds {
    guard ((boardId == BoardIds.STREAMING_BOARD) || (boardId == BoardIds.PLAYBACK_FILE_BOARD)) else {
        return boardId
    }
    if let otherInfoInt = Int32(params.other_info) {
        if let masterBoardId = BoardIds(rawValue: otherInfoInt) {
            return masterBoardId
        }
    }
    throw BrainFlowException ("need to set params.otherInfo to master board id",
                              BrainFlowExitCodes.INVALID_ARGUMENTS_ERROR)
}

struct BoardShim {
    let boardId: BoardIds
    let masterBoardId: BoardIds
    let bfParams: BrainFlowInputParams
    private let jsonBrainFlowInputParams: [CChar]
    
    init (_ boardId: BoardIds, _ params: BrainFlowInputParams) throws {
        self.boardId = boardId
        self.bfParams = params
        self.masterBoardId = try getMasterBoardID(boardId: boardId, params: params)
        self.jsonBrainFlowInputParams = params.json().cString(using: String.Encoding.utf8)!
    }
    
    /**
     * prepare steaming session, allocate resources
     */
    func prepareSession() throws {
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = prepare_session(boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to prepare session", errorCode: result)
    }
    
    /**
     * Get Board Id, can be different than provided (playback or streaming board)
     */
    func getBoardId () throws -> BoardIds {
        return self.masterBoardId
    }
        
    /**
     * send string to a board, use this method carefully and only if you understand what you are doing
     */
    func configBoard (_ config: String) throws -> String {
        var responseLen: Int32 = 0
        var response = [CChar](repeating: CChar(0), count: 4096)
        var cConfig = config.cString(using: String.Encoding.utf8)!
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = config_board (&cConfig, &response, &responseLen, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "Error in config_board", errorCode: result)

        return response.toString(responseLen)
    }

    /**
     * start streaming thread, store data in internal ringbuffer and stream them
     * from brainflow at the same time
     *
     * @param buffer_size     size of internal ringbuffer
     *
     * @param streamer_params supported vals: "file://%file_name%:w",
     *                        "file://%file_name%:a",
     *                        "streaming_board://%multicast_group_ip%:%port%". Range
     *                        for multicast addresses is from "224.0.0.0" to
     *                        "239.255.255.255"
     */
    func startStream (bufferSize: Int32, streamerParams: String) throws {
        var cStreamerParams = streamerParams.cString(using: String.Encoding.utf8)!
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = start_stream (bufferSize, &cStreamerParams, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to start stream", errorCode: result)
    }
   
    /**
     * start streaming thread, store data in internal ringbuffer
     */
    func startStream () throws {
        try startStream (bufferSize: 450000, streamerParams: "")
    }

    /**
     * start streaming thread, store data in internal ringbuffer
     */
    func startStream (bufferSize: Int32) throws {
        try startStream (bufferSize: bufferSize, streamerParams: "")
    }
    
    /**
     * stop streaming thread
     */
    func stopStream () throws {
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = stop_stream (boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "Error in stop_stream", errorCode: result)
    }
    
    /**
     * release all resources
     */
    func releaseSession () throws {
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = release_session (boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to release session", errorCode: result)
    }
    
    /**
     * get number of packages in ringbuffer
     */
    func getBoardDataCount () throws -> Int32 {
        var dataCount: Int32 = 0
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = get_board_data_count (&dataCount, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to get board data count", errorCode: result)
        
        return dataCount
    }

    /**
     * insert marker to data stream
     */
    func insertMarker (value: Double) throws {
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = insert_marker (value, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "Error in insert_marker", errorCode: result)
    }
    
    /**
     * check session status
     */
    func isPrepared () throws -> Bool  {
        var intPrepared: Int32 = 0
        var jsonBFParams = self.jsonBrainFlowInputParams
        let result = is_prepared (&intPrepared, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to check session", errorCode: result)
        
        return NSNumber(value: intPrepared).boolValue
    }

    /**
     * get latest collected data, can return less than "num_samples", doesnt flush
     * it from ringbuffer
     */
    func getCurrentBoardData (_ numSamples: Int32) throws -> [Double]
    {
        var numRows: Int32 = 0
        var currentSize: Int32 = 0
        var jsonBFParams = self.jsonBrainFlowInputParams

        do {
            numRows = try getNumRows (boardId: getBoardId()) }
        catch {
            throw error
        }
            
        var buffer = [Double](repeating: 0.0, count: Int(numSamples * numRows))
        let result = get_current_board_data (numSamples, &buffer, &currentSize, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "Error in get_current_board_data", errorCode: result)
        
        return buffer
    }

    /**
     * get all data from ringbuffer and flush it
     */
    func getBoardData () throws -> [[Double]] {
        var size: Int32 = 0
        var numRows: Int32 = 0
        var jsonBFParams = self.jsonBrainFlowInputParams

        do {
            size = try getBoardDataCount()
            guard size > 0 else {
                return [[Double]]()
            }
            numRows = try getNumRows (boardId: getBoardId()) }
        catch {
            throw error
        }
        
        var buffer = [Double](repeating: 0.0, count: Int(size * numRows))
        
        let result = get_board_data (size, &buffer, boardId.rawValue, &jsonBFParams)
        try checkErrorCode(errorMsg: "failed to get board data", errorCode: result)

        return buffer.matrix2D(rowLength: Int(size))
    }
}

File: brainflow-rust/crates/brainflow/src/board_shim.rs
//! BoardShim - High-level board interface.

use alloc::boxed::Box;
use alloc::string::String;
use alloc::vec::Vec;

use brainflow_boards::{
    Board, BoardController, BoardDescription, BoardId, BoardParams, Preset,
    board::BoardState,
    synthetic::SyntheticBoard,
};
use brainflow_sys::error::{Error, ErrorCode, Result};

/// High-level interface for board communication.
///
/// This matches the C++ BoardShim API.
pub struct BoardShim {
    board_id: BoardId,
    params: BoardParams,
    board: Option<Box<dyn Board>>,
}

impl BoardShim {
    /// Create a new BoardShim.
    pub fn new(board_id: BoardId, params: BoardParams) -> Result<Self> {
        Ok(Self {
            board_id,
            params,
            board: None,
        })
    }

    /// Get the board ID.
    pub fn board_id(&self) -> BoardId {
        self.board_id
    }

    /// Prepare the session (initialize hardware).
    pub fn prepare_session(&mut self) -> Result<()> {
        if self.board.is_some() {
            return Err(Error::new(ErrorCode::PortAlreadyOpen));
        }

        let mut board: Box<dyn Board> = match self.board_id {
            BoardId::Synthetic => Box::new(SyntheticBoard::new()),
            _ => {
                return Err(Error::with_message(
                    ErrorCode::UnsupportedBoard,
                    "Board not yet implemented",
                ))
            }
        };

        board.prepare_session()?;
        self.board = Some(board);
        Ok(())
    }

    /// Release the session (cleanup and disconnect).
    pub fn release_session(&mut self) -> Result<()> {
        if let Some(mut board) = self.board.take() {
            board.release_session()
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Start data streaming.
    ///
    /// # Arguments
    /// * `buffer_size` - Size of the ring buffer in samples.
    pub fn start_stream(&mut self, buffer_size: usize) -> Result<()> {
        if let Some(ref mut board) = self.board {
            board.start_stream(buffer_size)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Stop data streaming.
    pub fn stop_stream(&mut self) -> Result<()> {
        if let Some(ref mut board) = self.board {
            board.stop_stream()
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Get current board data without removing it from the buffer.
    pub fn get_current_board_data(&self, num_samples: usize) -> Result<Vec<f64>> {
        self.get_current_board_data_preset(num_samples, Preset::Default)
    }

    /// Get current board data for a specific preset.
    pub fn get_current_board_data_preset(
        &self,
        num_samples: usize,
        preset: Preset,
    ) -> Result<Vec<f64>> {
        if let Some(ref board) = self.board {
            board.get_current_board_data(num_samples, preset)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Get board data and remove it from the buffer.
    pub fn get_board_data(&self, num_samples: Option<usize>) -> Result<Vec<f64>> {
        self.get_board_data_preset(num_samples, Preset::Default)
    }

    /// Get board data for a specific preset.
    pub fn get_board_data_preset(
        &self,
        num_samples: Option<usize>,
        preset: Preset,
    ) -> Result<Vec<f64>> {
        if let Some(ref board) = self.board {
            board.get_board_data(num_samples, preset)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Get the number of samples in the buffer.
    pub fn get_board_data_count(&self) -> Result<usize> {
        self.get_board_data_count_preset(Preset::Default)
    }

    /// Get the number of samples for a specific preset.
    pub fn get_board_data_count_preset(&self, preset: Preset) -> Result<usize> {
        if let Some(ref board) = self.board {
            board.get_board_data_count(preset)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Insert a marker into the data stream.
    pub fn insert_marker(&mut self, value: f64) -> Result<()> {
        self.insert_marker_preset(value, Preset::Default)
    }

    /// Insert a marker for a specific preset.
    pub fn insert_marker_preset(&mut self, value: f64, preset: Preset) -> Result<()> {
        if let Some(ref mut board) = self.board {
            board.insert_marker(value, preset)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Send configuration to the board.
    pub fn config_board(&mut self, config: &str) -> Result<String> {
        if let Some(ref mut board) = self.board {
            board.config_board(config)
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Check if the session is prepared.
    pub fn is_prepared(&self) -> bool {
        self.board.as_ref().map_or(false, |b| b.is_prepared())
    }

    /// Check if streaming is active.
    pub fn is_streaming(&self) -> bool {
        self.board.as_ref().map_or(false, |b| b.is_streaming())
    }

    /// Get the board description.
    pub fn get_board_description(&self) -> Result<BoardDescription> {
        self.get_board_description_preset(Preset::Default)
    }

    /// Get the board description for a specific preset.
    pub fn get_board_description_preset(&self, preset: Preset) -> Result<BoardDescription> {
        if let Some(ref board) = self.board {
            Ok(board.description(preset).clone())
        } else {
            Err(Error::new(ErrorCode::BoardNotCreated))
        }
    }

    /// Get the sampling rate.
    pub fn get_sampling_rate(&self) -> Result<u32> {
        Ok(self.get_board_description()?.sampling_rate)
    }

    /// Get EEG channel indices.
    pub fn get_eeg_channels(&self) -> Result<Vec<usize>> {
        Ok(self.get_board_description()?.eeg_channels)
    }

    /// Get accelerometer channel indices.
    pub fn get_accel_channels(&self) -> Result<Vec<usize>> {
        Ok(self.get_board_description()?.accel_channels)
    }

    /// Get gyroscope channel indices.
    pub fn get_gyro_channels(&self) -> Result<Vec<usize>> {
        Ok(self.get_board_description()?.gyro_channels)
    }

    /// Get the timestamp channel index.
    pub fn get_timestamp_channel(&self) -> Result<Option<usize>> {
        Ok(self.get_board_description()?.timestamp_channel)
    }

    /// Get the marker channel index.
    pub fn get_marker_channel(&self) -> Result<Option<usize>> {
        Ok(self.get_board_description()?.marker_channel)
    }

    /// Get the package number channel index.
    pub fn get_package_num_channel(&self) -> Result<Option<usize>> {
        Ok(self.get_board_description()?.package_num_channel)
    }

    /// Get the device name.
    pub fn get_device_name(&self) -> Result<String> {
        Ok(self.get_board_description()?.name)
    }
}

impl Drop for BoardShim {
    fn drop(&mut self) {
        // Best-effort cleanup
        let _ = self.release_session();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_board_shim_lifecycle() {
        let mut board = BoardShim::new(BoardId::Synthetic, BoardParams::new()).unwrap();

        assert!(!board.is_prepared());

        board.prepare_session().unwrap();
        assert!(board.is_prepared());

        board.start_stream(1000).unwrap();
        assert!(board.is_streaming());

        let count = board.get_board_data_count().unwrap();
        assert!(count > 0);

        board.stop_stream().unwrap();
        assert!(!board.is_streaming());

        board.release_session().unwrap();
        assert!(!board.is_prepared());
    }
}
